tests/data/wasm/multifaceted/src/lib.rs

```diff
@@ -0,0 +1,125 @@
+//! VRL Announcement Example
+//!
+//! A WASM implementation of syslog parsing for benchmarks
+
+#![deny(improper_ctypes)]
+use serde_json::Value;
+use std::collections::HashMap;
+use std::convert::TryInto;
+use vector_wasm::{hostcall, Registration, Role};
+
+/// Perform one time initialization and registration.
+///
+/// During this time Vector and the plugin can validate that they can indeed work together,
+/// do any one-time initialization, or validate configuration settings.
+///
+/// It's required that the plugin call [`vector_wasm::Registration::register`] before returning.
+#[no_mangle]
+pub extern "C" fn init() {
+    // Vector provides you with a [`vector_wasm::WasmModuleConfig`] to validate for yourself.
+    let config = hostcall::config().unwrap();
+    assert_eq!(config.role, Role::Transform);
+
+    // Finally, pass Vector a `vector_wasm::Registration`
+    Registration::transform().register().unwrap();
+}
+
+/// Process data starting from a given point in memory to another point.
+///
+/// It's not necessary for the plugin to actually read, or parse this data.
+///
+/// Call [`vector_wasm::hostcall::emit`] to emit a message out.
+///
+/// # Returns
+///
+/// This function should return a hint of the number of emitted messages.
+#[no_mangle]
+pub extern "C" fn process(data: u32, length: u32) -> u32 {
+    // Vector allocates a chunk of memory through the hostcall interface.
+    // You can view the data as a slice of bytes.
+    let data = unsafe {
+        std::ptr::slice_from_raw_parts_mut(data as *mut u8, length.try_into().unwrap())
+            .as_mut()
+            .unwrap()
+    };
+
+    // For now, Vector uses simple JSON serialization to simplify bindings for other languages.
+    //
+    // **Please note that WASM support is still unstable!**
+    //
+    // We expect to alter this format in the future after some event data model improvements.
+    let mut event: HashMap<String, Value> = serde_json::from_slice(data).unwrap();
+
+    // The following is equivalent to the remap script:
+    // . = parse_syslog!(.message)
+
+    let message = event.remove("message");
+
+    let message = message
+        .as_ref()
+        .map(|v| v.as_str().unwrap_or_default())
+        .unwrap_or_default();
+
+    let message = message.to_lowercase();
```

Comment:
I think we should aim to do the transformations here in the same order as the remap script to ensure parity.

Reply:
🤔 These are operations still seem to be happening in a different order than the equivalent remap script.

Reply:
I adjusted this in bd2acbf . I came around to the fact that they don't need to be the exact same ordering as it should be comparing the best implementation from each.